File: leetcode-blind75/isBalanced.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

var isBalanced = function(root) {

    //helper function 
    var dfs = function(root) {
        if (!root) {
            return [true, 0];
        }

        let left = dfs(root.left);
        let right = dfs(root.right);
        let balanced;

        if (left[0] == true && right[0] == true && Math.abs(left[1] - right[1]) <= 1) {
            balanced = true;
        } else {
            balanced = false;
        }
        return [balanced, 1+ Math.max(left[1], right[1])];
    } 
    return dfs(root)[0]
};

//test cases

console.log(
    isBalanced([3,9,20,null,null,15,7]),//true
    isBalanced([1,2,2,3,3,null,null,4,4]),//true
    isBalanced([2]),//false
    isBalanced([]),//true
    isBalanced([4, 5, 6, null, null]),//true
    isBalanced([4, 3, null, 2, 1]),//false
);

File: leetcode/isBalanced.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsBalanced(TreeNode root) {
        if (root == null)
        {
            return true;
        }
        
        int leftHeight = getHeight(root.left);
        int rightHeight = getHeight(root.right);
        if (Math.Abs(leftHeight - rightHeight) > 1) {
            return false;
        }
        return IsBalanced(root.left) && IsBalanced(root.right);
    }
    
    private int getHeight(TreeNode node) {
        if (node == null) 
        {
            return 0;
        }
        return 1 + Math.Max(getHeight(node.left), getHeight(node.right));
    }
}